### ISSUE TITLE
Timing issues with :editsrc

### ISSUE THREAD

**Issue by rmull** [conformal/xombrero#27]
The active tab does not redraw if I edit the tmp file (obtained from :editsrc) with the default setting of 100ms (sent to g_timeout_add to trigger open_external_editor_cb()).

If I change this value to 500ms, then the page does in fact redraw. I have a slow computer, though I'm not sure if that's related.

If I make a number of edits to the tmp file, the browser eventually segfaults. I suspect this is timing-related. This is the backtrace:

#0  0x00007ffff61f0b8a in webkit_web_data_source_get_encoding () from /usr/lib/libwebkitgtk-3.0.so.0
#1  0x000000000041198d in edit_src_cb (
    contents=0xd74bb0 "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.01 Transitional//EN\"\n        \"http://www.w3.org/TR/html4/loose.dtd\">\n        <html>\n        <head><title>Wylcome to LWN.net [LWN.net]</title>\n        <meta HT"..., 
    data=0xcd4f20) at ../externaleditor.c:249
#2  0x0000000000411acc in update_contents (args=args@entry=0xa8ba50) at ../externaleditor.c:79
#3  0x0000000000411b78 in open_external_editor_cb (data=0xa8ba50) at ../externaleditor.c:142
#4  0x00007ffff335124b in ?? () from /usr/lib/libglib-2.0.so.0
#5  0x00007ffff3350645 in g_main_context_dispatch () from /usr/lib/libglib-2.0.so.0
#6  0x00007ffff3350988 in ?? () from /usr/lib/libglib-2.0.so.0
#7  0x00007ffff3350dea in g_main_loop_run () from /usr/lib/libglib-2.0.so.0
#8  0x00007ffff5873585 in gtk_main () from /usr/lib/libgtk-3.so.0
#9  0x000000000040d0d9 in main (argc=0, argv=0x7fffffffeb60) at ../xombrero.c:8877

A segfault is not the only issue. After some edits I might get the following message printed to the console:

** (xombrero:5723): CRITICAL **: const gchar* webkit_web_data_source_get_encoding(WebKitWebDataSource*): assertion 'WEBKIT_IS_WEB_DATA_SOURCE(webDataSource)' failed

**Comment by marcopeereboom** [comment]
Does this still happen?

**Comment by rmull** [comment]
Hi Marco,

I haven't used xombrero for a while, so I grabbed the latest copy and
rebuilt it. First, it looks like the syntax for the conf file
external_editor directive changed, so I fixed up my old config to open a
vim in a new urxvt terminal window. At that point, I verified that the tmp
file was created and the callback started polling it for changes. When I
edited the file and saved the change (without exiting the editor) it seemed
like the callback stopped firing, and the tab never updated with the
written changes.

I need to spend a little more time with it to figure out what's going on,
but I just wanted to let you know that I'm not ignoring you :)

**Comment by rmull** [comment]
Okay, so the above issue is due to my vim configuration, where the backup save procedure moves the original file to a backup location and writes a new file in its place. This causes the inode to change, and so the callback loses track of the file it's watching. Not sure it's xombrero's responsibility to handle this - probably not.

If I add "set backupcopy=yes" to my .vimrc, the problem is resolved.

I don't have the slow computer that I used to be able to use to reliably trigger the reported segfault, but I tested on another fairly slow computer and was unable to trigger it. I do still get the assertion printed to the console every time I make an edit but it doesn't seem to be preventing the changes from loading and the segfault seems resolved.

I think this can be closed.
